src/common/webrtc/api.rs

```diff
@@ -0,0 +1,463 @@
+#![allow(dead_code)]
+use std::{
+    fmt::Debug,
+    io::{self, Cursor},
+    net::Ipv4Addr,
+    rc::Rc,
+    sync::{atomic::AtomicBool, Arc},
+    time::Duration,
+};
+
+#[cfg(feature = "esp32")]
+use crate::esp32::exec::Esp32Executor;
+
+#[cfg(feature = "native")]
+use crate::native::exec::NativeExecutor;
+use crate::{
+    common::grpc_client::{GrpcMessageSender, GrpcMessageStream},
+    proto::rpc::webrtc::v1::{
+        answer_request, answer_response, AnswerRequest, AnswerResponse, AnswerResponseDoneStage,
+        AnswerResponseInitStage, AnswerResponseUpdateStage, IceCandidate,
+    },
+};
+
+use base64::{engine::general_purpose, Engine};
+use futures_lite::{future::block_on, StreamExt};
+use prost::{DecodeError, EncodeError};
+use sdp::{
+    description::{
+        common::{Address, ConnectionInformation},
+        media::{MediaName, RangedPort},
+    },
+    MediaDescription, SessionDescription,
+};
+use serde::{Deserialize, Serialize};
+use smol::net::UdpSocket;
+use smol_timeout::TimeoutExt;
+use thiserror::Error;
+
+use super::{
+    candidates::Candidate,
+    certificate::Certificate,
+    dtls::DtlsConnector,
+    ice::{ICEAgent, ICECredentials},
+    io::WebRTCTransport,
+    sctp::{Channel, SctpProto},
+};
+
+#[derive(Error, Debug)]
+pub enum WebRTCError {
+    #[error("signaling server disconnected")]
+    SignalingDisconnected(),
+    #[error("invalid SDP offer")]
+    InvalidSDPOffer(String),
+    #[error("invalid signaling request")]
+    InvalidSignalingRequest,
+    #[error("can't marshal answer")]
+    AnswerMarshalError(#[from] serde_json::Error),
+    #[error("signaling error")]
+    SignalingError(String),
+    #[error("data channel error")]
+    DataChannelOpenError(),
+    #[error("webrtc io error")]
+    IoError(#[from] io::Error),
+    #[error("webrtc grpc message decode error")]
+    GrpcDecodeError(#[from] DecodeError),
+    #[error("webrtc grpc message encode error")]
+    GprcEncodeError(#[from] EncodeError),
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+    #[error(transparent)]
+    DtlsError(#[from] Box<dyn std::error::Error + Send + Sync>),
```

Comment:
DtlsError needs to be compatible with both esp32DtlsError and OpenSSLError there they need to be reduced to the Error trait which they both implement. Furthermore this avoid copying the error struct into the stack when bubbling them up through the call stack. 
`Send` and `Sync` are both marker traits which when added to a type tells the compiler that it can be shared across different threads and be accessed concurrently